### PR TITLE
KAFKA-8729: Add upgrade docs for KIP-467 on augmented produce response

### PR DESCRIPTION
**Pull request by guozhangwang**
Add a paragraph explaining the producer caller's expected behavior change on record validation failure scenarios that are improved by KIP-467.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@tuvtran @rayokota @hachikuji for reviews.

**Review by tuvtran — COMMENTED** [review]
LGTM overall! Should we also talk about the 4 new metrics here?

**Comment by guozhangwang** [comment]
I plan to add them in another PR where I will also add for other KIPs.

File: docs/upgrade.html [hunk 1]
```diff
@@ -45,6 +45,10 @@ <h5><a id="upgrade_240_notable" href="#upgrade_240_notable">Notable changes in 2
         The old overloaded functions are deprecated and we would recommend users to make their code changes to leverage the new methods (details
         can be found in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-520%3A+Add+overloaded+Consumer%23committed+for+batching+partitions">KIP-520</a>).
     </li>
+    <li>We've introduced a new <code>INVALID_RECORD</code> error in the produce response to distinguish from the <code>CORRUPT_MESSAGE</code> error. Previously both scenarios
```
Comment (hachikuji) [inline docs/upgrade.html]: Can you add a comment about the change in behavior for the producer? I think most users won't know what this means.

**Comment by guozhangwang** [comment]
@hachikuji Could you take another look? Since in 2.4 we did not implement the retry logic I would keep the doc side aligned correspondingly; if in 2.5 the retry logic is implemented we can update the docs again.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 2]
```diff
@@ -317,7 +317,9 @@
             NoReassignmentInProgressException::new),
     GROUP_SUBSCRIBED_TO_TOPIC(86, "Deleting offsets of a topic is forbidden while the consumer group is actively subscribed to it.",
         GroupSubscribedToTopicException::new),
-    INVALID_RECORD(87, "This record has failed the validation on broker and hence be rejected.", InvalidRecordException::new);
+    INVALID_RECORD(87, "This record has failed the validation on broker and hence be rejected.", InvalidRecordException::new),
+    PURGE_IN_QUEUE(88, "This record has failed because other records from the same batch has been rejected on " +
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: I think we decided not to do this for now, right?

File: docs/upgrade.html [hunk 3]
```diff
@@ -87,6 +87,13 @@ <h5><a id="upgrade_240_notable" href="#upgrade_240_notable">Notable changes in 2
...
         can be found in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-520%3A+Add+overloaded+Consumer%23committed+for+batching+partitions">KIP-520</a>).
     </li>
+    <li>We've introduced a new <code>INVALID_RECORD</code> error in the produce response to distinguish from the <code>CORRUPT_MESSAGE</code> error.
+        To be more concrete, previously when a batch of records were sent as part of a single request to the broker and one or more of the records failed
+        the validation due to various causes (mismatch magic bytes, crc checksum errors, null key for log compacted topics, etc), the whole batch would be rejected
+        with the same and misleading <code>CORRUPT_MESSAGE</code>, and the caller of the producer client would see the corresponding exception from either
+        the future object of <code>RecordMetadata</code> returned from the <code>send</code> call as well as in the <code>Callback#onCompletion(RecordMetadata metadata, Exception exception)</code>
+        Now with the new error code and improved error messages of the exception, producer callers would be better notified about the root cause why their sent records were failed.
```
Comment (hachikuji) [inline docs/upgrade.html]: nit: "would be better notified" -> "will be better informed"?

**Review by hachikuji — APPROVED** [review]
Two comments. LGTM once addressed.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: My bad, will remove.

File: docs/upgrade.html (hunk 3 above)
Comment (guozhangwang) [reply docs/upgrade.html]: Ack.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.4.

**Pull request merged**